File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        char_set = set()
        count = 0
        l = 0

        for r in range(len(s)):

            while s[r] in char_set:
                char_set.remove(s[l])
                l+=1

            char_set.add(s[r])
            count = max(count, r - l + 1)

        return count

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        Set<Character> set = new HashSet<>();
        int count = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            char current = s.charAt(right);
            while (set.contains(current)) {
                set.remove(s.charAt(left));
                left++;
            }
            set.add(current);
            count = Math.max(count, right - left + 1);
        }
        return count;
    }
}